pkg/utils/etcdutil/etcdutil.go

```diff
@@ -350,3 +353,240 @@ func InitOrGetClusterID(c *clientv3.Client, key string) (uint64, error) {
+	defer ticker.Stop()
+
+	for i := 0; i < lw.loadRetryTimes; i++ {
+		watchStartRevision, err = lw.load()
+		failpoint.Inject("loadTemporaryFail", func(val failpoint.Value) {
+			if maxFailTimes, ok := val.(int); ok && i < maxFailTimes {
+				err = errors.New("fail to read from etcd")
+				failpoint.Continue()
+			}
+		})
+		if err == nil {
+			if deadline, ok := ctx.Deadline(); ok && deadline.After(time.Now()) {
+				break
+			}
+		}
+		select {
+		case <-ctx.Done():
+			lw.isLoadedCh <- errors.Errorf("ctx is done before load data from etcd")
+			return watchStartRevision
+		case <-ticker.C:
+		}
+	}
+	if err != nil {
+		log.Warn("watched key doesn't exist in watch loop when loading", zap.String("name", lw.name), zap.String("key", lw.key), zap.Error(err))
+	}
+	lw.isLoadedCh <- err
+	return watchStartRevision
+}
+
+func (lw *LoopWatcher) watch(ctx context.Context, revision int64) (nextRevision int64, err error) {
+	watcher := clientv3.NewWatcher(lw.client)
+	defer watcher.Close()
+
+	for {
+	WatchChan:
+		opts := append(lw.opts, clientv3.WithRev(revision))
+		watchChan := watcher.Watch(ctx, lw.key, opts...)
+		select {
+		case <-ctx.Done():
+			return revision, nil
+		case <-lw.forceLoadCh:
+			revision, err = lw.load()
+			if err != nil {
+				log.Warn("force load key failed in watch loop", zap.String("name", lw.name),
+					zap.String("key", lw.key), zap.Error(err))
+			}
+			goto WatchChan
+		case wresp := <-watchChan:
+			if wresp.CompactRevision != 0 {
+				log.Warn("required revision has been compacted, use the compact revision in watch loop",
+					zap.Int64("required-revision", revision),
+					zap.Int64("compact-revision", wresp.CompactRevision))
+				revision = wresp.CompactRevision
+				goto WatchChan
+			} else if wresp.Err() != nil { // wresp.Err() contains CompactRevision not equal to 0
+				log.Error("watcher is canceled in watch loop",
+					zap.Int64("revision", revision),
+					errs.ZapError(errs.ErrEtcdWatcherCancel, wresp.Err()))
+				return revision, wresp.Err()
+			}
+			for _, event := range wresp.Events {
+				switch event.Type {
+				case clientv3.EventTypePut:
+					if err := lw.putFn(event.Kv); err != nil {
+						log.Error("put failed in watch loop", zap.String("name", lw.name),
+							zap.String("key", lw.key), zap.Error(err))
+					}
+				case clientv3.EventTypeDelete:
+					if err := lw.deleteFn(event.Kv); err != nil {
+						log.Error("delete failed in watch loop", zap.String("name", lw.name),
+							zap.String("key", lw.key), zap.Error(err))
+					}
+				}
+			}
+			if err := lw.postEventFn(); err != nil {
+				log.Error("run post event failed in watch loop", zap.String("name", lw.name),
+					zap.String("key", lw.key), zap.Error(err))
+			}
+			revision = wresp.Header.Revision + 1
+		}
+	}
+}
+
+func (lw *LoopWatcher) load() (nextRevision int64, err error) {
+	startKey := lw.key
+	limit := lw.loadBatchSize
+	if limit <= 0 {
+		limit = defaultLoadBatchSize
```

Comment:
defaultLoadBatchSize is only needed in some specific scenarios?

Reply:
when watch a lot of key